[CALCITE-6239, CALCITE-6263] Add PostGIS Dialect to Apache Calcite

bchapuis opened the pull request:
This PR introduces a new PostgisSqlDialect to Apache Calcite, providing initial support for PostGIS functions. It primarily addresses [CALCITE-6239](https://issues.apache.org/jira/browse/CALCITE-6239) and follows the architectural approach proposed by @julianhyde in [CALCITE-6263](https://issues.apache.org/jira/browse/CALCITE-6263).

**Summary of the changes:**
- New Dialect: PostgisSqlDialect class extends PostgresqlSqlDialect, specializing SQL generation for PostGIS functions.
- ST_ Function Support: Adds SQL translation for selected ST_ prefixed spatial functions (e.g., ST_Contains, ST_Within, etc.), enabling Calcite to generate syntactically correct PostGIS SQL from relational algebra expressions using these functions.
- Function Mapping: Introduced mapping from Calcite’s standard operator representations to PostGIS-compatible SQL functions.
- Test Coverage: Added unit tests to validate the translation logic for supported spatial functions under the new dialect.

**Coupling with Java Topology Suite (JTS):**
The main concern regarding this PR was about avoiding excessive coupling between Calcite’s core and third-party spatial libraries like JTS as described in [CALCITE-6263](https://issues.apache.org/jira/browse/CALCITE-6263). We now have a Built-in GEOMETRY Type independant from JTS. The coupling with JTS occurs in `JavaTypeFactoryImpl` which is better than in `RelDataTypeFactoryImpl`.

bchapuis committed:
[CALCITE-6239] Add a postgis dialect that supports ST functions

YiwenWu commented on core/src/main/java/org/apache/calcite/sql/dialect/PostgisSqlDialect.java:
```diff
@@ -0,0 +1,51 @@
+    super(context);
+  }
+
+  @Override public void unparseCall(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
```
Since PostGIS is an extension of PostgreSQL, Can these changes be processed directly in PostgresqlSqlDialect?

bchapuis replied:
Yes they could, I also had mixed feeling about extending the postgresql dialect. My current reasoning comes from the fact that postgis is an extension that must be installed on the server and then activated in the database with `CREATE EXTENSION postgis;`. Having a separate dialect makes things more explicit in calcite and isolates postgis specific customizations in a dedicated `unparseCall` method.

YiwenWu replied:
It would be clearer to have an independent Dialect, but I am worried that it will introduce a maintenance burden. 
At present, PostGIS does not have many unique functions.

bchapuis replied:
ST_UNION is the main case I had in mind that required to introduce a new function name (ST_UNARYUNION) in calcite. However, some other functions, such as ST_DUMP, may require a similar treatment. I guess that the list will grow as the dialect gets more systematically tested against postgis.

julianhyde replied:
Simplest thing seems to be to make Postgis dialect a subclass of Postgres dialect. There are many similarities, but also many differences.

YiwenWu commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -7876,6 +7878,26 @@ private void checkLiteral2(String expression, String expected) {
+   */
+  @Test void testPostgisStSrid() {
+    String query = "select st_srid(\"point\") FROM \"points\"";
+    String expectedPostgis = "SELECT \"ST_SRID\"(\"point\")\nFROM \"GEO\".\"points\"";
```
Some UT tests failed

bchapuis replied:
I should probably have created a draft PR.

JiajunBernoulli commented on core/src/main/java/org/apache/calcite/sql/SqlDialectFactoryImpl.java:
```diff
@@ -222,6 +223,8 @@ public class SqlDialectFactoryImpl implements SqlDialectFactory {
       return PhoenixSqlDialect.DEFAULT;
     case POSTGRESQL:
       return PostgresqlSqlDialect.DEFAULT;
+    case POSTGIS:
```
Is `POSTGIS` before `POSTGRESQL`?
> I is before R

bchapuis committed:
Fix the registration of the geometry schema

bchapuis committed:
Place postgis before postgres whenever it makes sense

macroguo-ghy commented:
> @YiwenWu @JiajunBernoulli What would be the best way to test these changes against a postgis database? Ideally, I'd like to add an integration test that executes queries against postgis (e.g. with testcontainers). But I havn't been able to find such tests in calcite.

I remember that testContainer has been used in redis apapter test.

JiajunBernoulli commented:
> @YiwenWu @JiajunBernoulli What would be the best way to test these changes against a postgis database? Ideally, I'd like to add an integration test that executes queries against postgis (e.g. with testcontainers). But I havn't been able to find such tests in calcite.

There are some integration tests for Druid: https://github.com/zabetak/calcite-druid-dataset
Here is CI config in calcite: https://github.com/apache/calcite/blob/2aabf210dc1918c6ca20e63b39661ff445535eb8/.github/workflows/main.yml#L440

bchapuis commented:
Thanks a lot for the pointer, I will adopt the same approach and implement some integration tests in a third-party repository for now.

bchapuis committed:
Add support for implicit type conversion from binary to geometry

bchapuis committed:
Add a geometry parser to bind jdbc geometries to calcite geometries

bchapuis committed:
Improve naming and fix minor issues

bchapuis committed:
Use lenient conformance

julianhyde commented on core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcSchema.java:
```diff
@@ -433,6 +433,12 @@ private static RelDataType sqlType(RelDataTypeFactory typeFactory, int dataType,
                 typeFactory.createSqlType(SqlTypeName.ANY), true);
       }
       return typeFactory.createArrayType(component, -1);
+    case OTHER:
```
On most databases the type string will be upper-case, "GEOMETRY".

bchapuis replied:
Yes, the use of OTHER with a "geometry" type name seems specific to the postgresql jdbc driver and postgis. I guess they did this to distinguish postgis geometry types from the regular geometry types implemented in postgresql. https://www.postgresql.org/docs/current/datatype-geometric.html

I will clarify this with a comment and try to investigate regular geometry data types a bit more.

julianhyde commented on core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcToEnumerableConverter.java:
```diff
@@ -283,6 +284,15 @@ private static void generateGet(EnumerableRelImplementor implementor,
       break;
+    case GEOMETRY:
+      source =
+          Expressions.call(
```
Add methods to BuiltInMethod. We don't want reflection errors at runtime.

julianhyde commented on core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:
```diff
@@ -1488,6 +1505,7 @@ public static SqlNode toSql(RexLiteral literal) {
       default:
         break;
       }
+
```
remove blank line

julianhyde commented on core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java:
```diff
@@ -39,9 +39,11 @@
 import com.google.common.collect.Interners;
 
 import org.checkerframework.checker.nullness.qual.Nullable;
+import org.locationtech.jts.geom.Geometry;
```
Do we really want this dependency here?

bchapuis replied:
I'm trying to wrap my head around this. From what I understand, calling a method of RelDataTypeFactoryImpl without JTS would trigger a class not found exception. As the use of JTS is currently circumscribed to a few spatial type classes, calcite works just fine without adding a dependency to JTS. Is that correct?

Given the popularity of JTS in the geospatial world, I wouldn't try to abstract it completely in JTS. An option could be to use reflection to dynamically load the Geometry class and provide some sort of fallback if the class is not in the class path.

```java
public static class Geometry { }

static Class<?> loadGeometryClass() {
    try {
        return Class.forName("org.locationtech.jts.geom.Geometry");
    } catch (ClassNotFoundException e) {
        return Geometry.class;
    }
}

private static final Map<Class, RelDataTypeFamily> CLASS_FAMILIES =
  ImmutableMap.<Class, RelDataTypeFamily>builder()
      .put(String.class, SqlTypeFamily.CHARACTER)
      // ...
      .put(loadGeometryClass(), SqlTypeFamily.GEO)
      .build();
```

What do you think?

julianhyde replied:
Can we discuss in jira? Increasing the coupling to JTS is an architectural change.

bchapuis replied:
Yes, here is the jira issue I just created:
https://issues.apache.org/jira/browse/CALCITE-6263

julianhyde commented on core/src/main/java/org/apache/calcite/sql/SqlDialect.java:
```diff
@@ -1382,6 +1384,7 @@ public enum DatabaseProduct {
     LUCIDDB("LucidDB", "\"", NullCollation.HIGH),
     INTERBASE("Interbase", null, NullCollation.HIGH),
     PHOENIX("Phoenix", "\"", NullCollation.HIGH),
+    POSTGIS("Postgis", "\"", NullCollation.HIGH),
```
Usual capitalization is PostGIS, I believe

julianhyde commented on core/src/main/java/org/apache/calcite/sql/SqlKind.java:
```diff
@@ -1140,27 +1140,447 @@ public enum SqlKind {
    * Hilbert space-filling curve. */
   HILBERT,
 
+  ST_MAKE_LINE,
```
Are these obsolete or not?

bchapuis replied:
I believe we can remove them. ST_MAKE_LINE is ST_MAKELINE and all results returned by search engines for ST_POINT3 are related to ST_POINT.

julianhyde commented on core/src/test/java/org/apache/calcite/sql/test/SqlTypeNameTest.java:
```diff
@@ -172,7 +173,7 @@ class SqlTypeNameTest {
     SqlTypeName tn =
         SqlTypeName.getNameForJdbcType(Types.OTHER);
-    assertEquals(null, tn, "OTHER did not map to null");
+    assertEquals(OTHER, tn, "OTHER did not map to null");
```
huh?

bchapuis replied:
I remember modifying this line and found the test cases unclear (and I forgot to adapt the error message). The javadoc of getNameForJdbcType states: "corresponding SqlTypeName, or null if the type is not known". As other is kind of a known type, does it makes sense to adapt the test accordingly? I will commit this fix for now.

```java
  @Test void testOther() {
    SqlTypeName tn =
        SqlTypeName.getNameForJdbcType(Types.OTHER);
    assertEquals(OTHER, tn, "OTHER did not map to OTHER");
  }
```

julianhyde commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeName.java:
```diff
@@ -264,6 +264,8 @@ public enum SqlTypeName {
           .put(Types.STRUCT, STRUCTURED)
           .put(Types.ARRAY, ARRAY)
+          .put(Types.OTHER, OTHER)
+
```
remove blank line

julianhyde commented on core/src/main/java/org/apache/calcite/sql/SqlKind.java:
```diff
@@ -1140,27 +1140,447 @@ public enum SqlKind {
+  ST_UNARYUNION,
+
+  /** The {@code ST_UnaryUnion} function. */
+  ST_UNION,
```
javadoc should be ST_Union

julianhyde commented on core/src/main/java/org/apache/calcite/sql/dialect/PostgisSqlDialect.java:
```diff
@@ -0,0 +1,53 @@
+
+  @Override public void unparseCall(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    if (call.getOperator().getName().equals("ST_UNARYUNION")) {
+      RelToSqlConverterUtil.specialOperatorByName("ST_UNION")
```
use SqlKind rather than name?

julianhyde commented on core/src/main/java/org/apache/calcite/sql/dialect/PostgisSqlDialect.java:
```diff
@@ -0,0 +1,53 @@
+
+  public static final SqlDialect DEFAULT = new PostgisSqlDialect(DEFAULT_CONTEXT);
+
+  /** Creates a PostgresqlSqlDialect. */
```
a few places you wrote 'Postgresql' where you meant 'PostGIS' or 'Postgis'.

julianhyde commented on core/src/main/java/org/apache/calcite/sql/dialect/PostgisSqlDialect.java:
```diff
@@ -0,0 +1,53 @@
+
+/**
+ * A <code>SqlDialect</code> implementation for the PostgreSQL database.
+ */
```
explain why it is a subclass of PostgresSQLDialect

julianhyde commented on core/src/main/java/org/apache/calcite/sql/dialect/PostgisGeometryDecoder.java:
```diff
@@ -0,0 +1,191 @@
+import java.util.Arrays;
+
+/**
+ * A decoder for the PostGIS geometry format.
```
This class seems like a utility class, not appropriate for the org.apache.calcite.sql.dialect package. I didn't see any unit tests, but it could use a few.

bchapuis replied:
ok, I will place it in org.apache.calcite.util and add some unit tests.

julianhyde reviewed (commented):
Can you change the summary to "In JDBC adapter, add a PostGIS dialect so that ST_ functions can be pushed down". It will make the release notes easier to understand.

jnh5y commented on core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:
```diff
@@ -312,7 +312,9 @@ private Expression getConvertExpression(
-
+      case BINARY:
+      case VARBINARY:
+        return Expressions.call(BuiltInMethod.ST_GEOM_FROM_EWKB.method, operand);
```
As a future consideration, there are other binary formats like Tiny WKB.  https://github.com/TWKB/Specification/blob/master/twkb.md

I suppose that may be handled by a specific cast.

bchapuis replied:
Thanks a lot for the pointer, it would definitely make sense to implement these functions:
https://postgis.net/docs/ST_AsTWKB.html

bchapuis committed:
Ignore casing when checking the type name

bchapuis committed:
Add a built in method to decode postgis geometries

This prevent reflection errors at runtime.

bchapuis committed:
Remove unnecessary comment

For now, let's assume that most databases behave as postgis.

bchapuis committed:
Use reflection to register the geometry type family

bchapuis committed:
Consistently use PostGIS instead of Posgis

bchapuis committed:
Remove unnecessary functions from SqlKind

bchapuis committed:
Document the extension of the PostgreSQL dialect by the PostGIS dialect

bchapuis committed:
Use SqlKind instead of Strings in the PostGIS dialect

bchapuis committed:
Fix the error message of the unit test for the OTHER type

bchapuis committed:
Format the code with autostyle

bchapuis committed:
Fix style and coding issues

bchapuis committed:
Fix the javadoc

bchapuis committed:
Move the PostgisGeometryDecoder into the util package

bchapuis committed:
Add test cases for the postgis geometry decoder

bchapuis committed:
Remove reference to removed ST functions

bchapuis committed:
Format the code

bchapuis committed:
Revert changes in the sql dialect

bchapuis committed:
Check null value in the postgis geometry decoder

bchapuis committed:
Add a lexical policy for PostgreSQL

bchapuis committed:
Add nullable annotation to geometry decoder

bchapuis committed:
Improve casting of decoded geometries

YiwenWu commented on core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcSchema.java:
```diff
@@ -433,6 +433,17 @@ private static RelDataType sqlType(RelDataTypeFactory typeFactory, int dataType,
                 typeFactory.createSqlType(SqlTypeName.ANY), true);
       }
       return typeFactory.createArrayType(component, -1);
+    case GEOMETRY:
```
According to the elements in `SqlTypeName#JDBC_TYPE_TO_NAME` , whether there are any cases that can be matched `GEOMETRY`?

YiwenWu commented on core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:
```diff
@@ -1445,6 +1447,21 @@ public static SqlNode toSql(RexLiteral literal) {
     }
+    case GEO: {
+      Geometry geom = castNonNull(literal.getValueAs(Geometry.class));
+      switch (typeName) {
```
I am not sure that typeName `CHAR` or `VARCHAR` can be obtained under `GEO`  family.

julianhyde committed:
Add annotation Parameter.sqlType

The annotation allows you to define the SQL type of a
parameter, for cases where it cannot be deduced automatically
from the Java type. This is especially useful for GEOMETRY
values, for which Calcite has no built-in Java type. (JTS
Geometry is widely used in Calcite, but it is an
implementation, and we do not want it to become the
specification.)

This commit is incomplete, but illustrates the general idea.

A related change is to add a 'family' parameter to
the `TypeFactory.createJavaType` method. It is nullable, and
if null, the type becomes its own family. In JavaType, we
should add a `family` field.

When this change is complete we should be able to remove JTS
Geometry from RelDataTypeFactoryImpl.CLASS_FAMILIES.

julianhyde committed:
More

bchapuis committed:
Adapt more functions to the geometry type parameter

bchapuis committed:
Remove the JTS type

mihaibudiu commented:
Will this be ready for 1.37?

bchapuis commented:
No, I hadn't enought time to work on this and the decoupling of the GEOMETRY type from the JTS Geometry class will require additional efforts.

https://issues.apache.org/jira/browse/CALCITE-6239

bchapuis committed:
Resolve conflicts

bchapuis committed:
Add support for Java type family in RelDataTypeFactory

bchapuis committed:
Revert change

bchapuis committed:
Fix test for OTHER

bchapuis committed:
Revert changes in type mapping

bchapuis committed:
Remove diff

bchapuis committed:
Revert some changes

bchapuis committed:
Revert some changes and make some cleaning

bchapuis committed:
Remove JTS
Geometry from RelDataTypeFactoryImpl.CLASS_FAMILIES

bchapuis committed:
Add JDBC_TYPE_TO_NAME record for geometries

mihaibudiu commented on core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactory.java:
```diff
@@ -65,6 +65,15 @@ public interface RelDataTypeFactory {
+   * Creates a type that corresponds to a Java class, with a given family.
+   *
+   * @param clazz the Java class used to define the type
+   * @param family The family of the type, or null to infer
```
but perhaps this PR is not the right place to have this conversation.

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeAssignmentRule.java:
```diff
@@ -193,6 +193,8 @@ private SqlTypeAssignmentRule(
     rule.add(SqlTypeName.GEOMETRY);
     rule.add(SqlTypeName.CHAR);
     rule.add(SqlTypeName.VARCHAR);
+    rule.add(SqlTypeName.BINARY);
```
these (and a few other similar changes) look like bug fixes that would be nice to have even before we can merge the full PR.

bchapuis replied:
I think some of these bug fixes have been addressed separately by Lucas Brenner in CALCITE-7073. I had a few merge conflicts when rebasing the current branch.

I'm not completely sure what else could be extracted into smaller PR/bug fixes and would prefer to squash merge this PR to move on ;)

At this point I'm not completely sure what else could be extracted into smaller PR/bug fixes. To keep things moving, I’d prefer to squash-merge, unless anyone has other suggestions.

mihaibudiu commented on core/src/test/java/org/apache/calcite/util/PostgisGeometryDecoderTest.java:
```diff
@@ -0,0 +1,124 @@
+        new WKTWriter().write(geometry));
+  }
+
+  @Test void decodeGeometryCollection() {
```
how about a few negative tests, where the decoding cannot happen?

mihaibudiu reviewed (approved):
This looks pretty nice to me; the PR is relatively small given how much it accomplishes.
But I am not familiar with this subject.
One thing I wonder about: if other dialects want to provide geography support, what do they need to do? We don't really want only Postgis to do that.

zabetak commented on core/src/main/java/org/apache/calcite/config/Lex.java:
```diff
@@ -63,6 +63,13 @@ public enum Lex {
+   * Unquoted identifiers are converted to lower-case; after which, identifiers
+   * are matched case-sensitively.
+   */
+  POSTGRESQL(Quoting.DOUBLE_QUOTE, Casing.TO_LOWER, Casing.UNCHANGED, false,
```
Where do we use this? How is it tested?

bchapuis replied:
I added a test cases for this in `LexCaseSensitiveTest` and in `LexEscapeTest`.

My integration tests are located here: https://github.com/bchapuis/calcite-postgis-tests

zabetak commented on core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcToEnumerableConverter.java:
```diff
@@ -323,6 +323,14 @@ private static void generateGet(EnumerableRelImplementor implementor,
               java.sql.Array.class);
       source = Expressions.call(BuiltInMethod.JDBC_ARRAY_TO_LIST.method, x);
       break;
+    case GEOMETRY:
```
Are there tests that cover this path?

zabetak replied:
We have usages of testcontainers in Calcite. Is there something that  prevents the integration tests to be committed as part of this PR?

bchapuis replied:
My integration tests are located here: https://github.com/bchapuis/calcite-postgis-tests

I looked into how Calcite uses Testcontainers, but it seems the support is currently limited to a few Redis tests. The JDBC tests don’t appear to leverage Testcontainers yet, which is why I set up external integration tests. Did I overlook anything?

zabetak reviewed (changes requested):
Thanks for advancing this work @bchapuis !

My main concerns are around breaking changes that could potentially be avoided mostly around the type factories and the additional coupling with `org.locationtech` that appears in some places such as the `SqlImplementor`.

Most of the remaining comments/questions are rather minor.

bchapuis commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java:
```diff
@@ -76,7 +76,7 @@ public enum SqlTypeFamily implements RelDataTypeFamily {
   CURSOR,
   COLUMN_LIST,
-  GEO,
+  GEOMETRY,
```
Maybe, I can revert it.

bchapuis commented on linq4j/src/main/java/org/apache/calcite/linq4j/function/Parameter.java:
```diff
@@ -81,4 +82,10 @@
    * is NULL, and therefore optional parameters must be nullable.
    */
   boolean optional() default false;
+
+  /** Returns the SQL type code.
+   *
+   * <p>Values are typically from {@link java.sql.Types}, for example
+   * {@link java.sql.Types#INTEGER}. */
+  int sqlType() default Types.OTHER;
```
This was introduced in the following commit:
https://github.com/apache/calcite/pull/3668/commits/da8494213c3810584a75064d5fadf52ac1f95d12

> Add annotation Parameter.sqlType
> The annotation allows you to define the SQL type of a
parameter, for cases where it cannot be deduced automatically
from the Java type. This is especially useful for GEOMETRY
values, for which Calcite has no built-in Java type. (JTS
Geometry is widely used in Calcite, but it is an
implementation, and we do not want it to become the
specification.)

bchapuis committed:
Improve isolation of postgis decoder tests

bchapuis commented on core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:
```diff
@@ -1511,6 +1513,21 @@ public static SqlNode toSql(RexLiteral literal) {
+        return SqlLiteral.createBinaryString(
+            SpatialTypeUtils.asWkbArray(geom), POS);
+      default:
+        return SqlLiteral.createNull(POS);
```
I used createUnknown for now.

bchapuis committed:
Add negative tests and improve exceptions consistency

bchapuis commented on core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java:
```diff
@@ -837,7 +837,7 @@ private static ColumnMetaData.AvaticaType avaticaType(JavaTypeFactory typeFactor
         return ColumnMetaData.struct(columns);
       case ExtraSqlTypes.GEOMETRY:
-        typeOrdinal = Types.VARCHAR;
+        typeOrdinal = Types.OTHER;
```
I checked the history.md file. Is it modified by the release manager? I'm tempted to say that this change is closer to a bugfix and has a relatively low impact as it matches what is done in other databases and drivers.

bchapuis committed:
Add default implementation for createJavaType(Class clazz)

bchapuis committed:
Add test for PostgreSQL lexical policy

bchapuis committed:
Fix formatting

bchapuis committed:
Remove multiple entries with same key

bchapuis committed:
Fix issues related to OTHER in PostgreSQL

bchapuis committed:
Add test cases for PostgreSQL lex

bchapuis committed:
Removed irrelevant changes

bchapuis committed:
Rename GEOMETRY to GEO to prevent breaking changes

bchapuis committed:
Clean commented code block used for code generation

bchapuis commented on core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java:
```diff
@@ -22,6 +22,7 @@
 import org.apache.calcite.linq4j.function.Deterministic;
 import org.apache.calcite.linq4j.function.Experimental;
 import org.apache.calcite.linq4j.function.Hints;
+import org.apache.calcite.linq4j.function.Parameter;
```
This is related to the approach introduced here:
https://github.com/apache/calcite/commit/7124c3ae3e5001b85dc6617554c170be49476f75

bchapuis commented on core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:
```diff
@@ -116,6 +117,7 @@
 
 import org.checkerframework.checker.initialization.qual.UnknownInitialization;
 import org.checkerframework.checker.nullness.qual.Nullable;
+import org.locationtech.jts.geom.Geometry;
```
Would this imply adding a wrapper in Calcite to reduce coupling to JTS? I’m still not convinced that the added complexity is worth it. JTS is almost a standard for spatial functions and operators. GEOS, which is used in Postgis, is actually a C++ port of JTS, and their communities overlap significantly. To put it another way,  I think JTS is almost to geometry what Guava is to immutable maps and lists ;)

bchapuis committed:
Fix test

bchapuis commented:
@zabetak Thanks a lot for your review. I think most of the comments have been addressed. Please let me know if you want to further discuss de decoupling of JTS.

bchapuis committed:
Change the default value for GEO

bchapuis committed:
Describe the breaking change affecting the next release

zabetak reviewed (commented):
Did another quick pass and left some comments/clarifications. Currently, I am pretty busy with other stuff so it's kinda hard to be very reactive on this PR. Feel free to move this forward if others are OK with it and I don't respond reasonably fast.
